source/utils.py

```diff
@@ -0,0 +1,278 @@
+"""General LoadInsight utilities
+
+Note: many of these utilities cannot use `config` during module load and class init
+because the running configuration has not been loaded `load_config()`
+"""
+import os
+import inspect
+import pandas as pd
+import json
+
+config = None
+def load_config(name=None):
+    global config
+    if config == None:
+        if name == None:
+            raise Exception("config name is not defined yet (did you forget to call utils.load_config first?)")
+        import importlib
+        config = importlib.import_module(name)
+    elif not name == None:
+        raise Exception("cannot load a different configuration (%s is already loaded)" % (config))
+    return config
+
+class safecache:
```

Comment:
what's the purpose of this cache?

Reply:
The purpose of the cache is
1. avoid reading data frames from disk that are already available elsewhere in memory
2. avoid writing data frames to disk that have already been written to disk
3. avoid having multiple copies of the same data frame in memory, unless necessary
4. maintain the consistency of local and remote copies of data frames.

Reply:
So yeah, I assumed 1-3. And I think generally it's a good idea, but especially when running pipelines in parallel, allowing so much to remain in memory (for such a long time, assuming hours scale and not minutes) can be extremely costly and may necessitate pretty large instances on AWS for no real good reason. If we can rely on the GC and ensure the I/O isn't too bad in the case of re-reading a DF, that may be a better approach. However, something we'd need to validate for sure.

As far as number 4, to me it seemed like that should be the responsibility of the Data Manager and not of a caching utility.

Anyways, just wanted to get your thoughts on this, thanks!

Reply:
Agreed re 4. The implementation is problematic. I’m still working on it and remove 4 would mean it basically works now because 1-3 are easy and effective. Let me know how you’d like to approach 4.

Reply:
I need to think about how to do this so that using the same name in different pipelines doesn't cause trouble when data is deleted when it is no longer used.